pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/GrpcBrokerRequestHandler.java

```diff
@@ -0,0 +1,198 @@
+      @Nullable BrokerRequest realtimeBrokerRequest, @Nullable Map<ServerInstance, List<String>> realtimeRoutingTable,
+      long timeoutMs, ServerStats serverStats, RequestStatistics requestStatistics)
+      throws Exception {
+    assert offlineBrokerRequest != null || realtimeBrokerRequest != null;
+
+    String rawTableName = TableNameBuilder.extractRawTableName(originalBrokerRequest.getQuerySource().getTableName());
+    Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> responseMap = new HashMap<>();
+    // Making request to a streaming server response.
+
+    if (offlineBrokerRequest != null) {
+      // to offline servers.
+      assert offlineRoutingTable != null;
+      Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> serverResponses =
+          streamingQueryToPinotServer(requestId, _brokerId, rawTableName, TableType.OFFLINE,
+              offlineBrokerRequest, offlineRoutingTable, timeoutMs, true, 1);
+      responseMap.putAll(serverResponses);
+    }
+    if (realtimeBrokerRequest != null) {
+      // to realtime servers.
+      assert realtimeRoutingTable != null;
+      Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> serverResponses =
+          streamingQueryToPinotServer(requestId, _brokerId, rawTableName, TableType.REALTIME,
+              realtimeBrokerRequest, realtimeRoutingTable, timeoutMs, true, 1);
+      responseMap.putAll(serverResponses);
+    }
+    BrokerResponseNative brokerResponse = _streamingReduceService.reduceOnStreamResponse(
+        originalBrokerRequest, responseMap, timeoutMs, _brokerMetrics);
+    return brokerResponse;
+  }
+
+  /**
+   * Query pinot server for data table.
+   */
+  public Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> streamingQueryToPinotServer(
+      final long requestId,
+      final String brokerHost,
+      final String rawTableName,
+      final TableType tableType,
+      BrokerRequest brokerRequest,
+      Map<ServerInstance, List<String>> routingTable,
+      long connectionTimeoutInMillis,
+      boolean ignoreEmptyResponses,
+      int pinotRetryCount) {
+    // Retries will all hit the same server because the routing decision has already been made by the pinot broker
+    Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> serverResponseMap = new HashMap<>();
+    for (Map.Entry<ServerInstance, List<String>> routingEntry : routingTable.entrySet()) {
+      ServerInstance serverInstance = routingEntry.getKey();
+      List<String> segments = routingEntry.getValue();
+      String serverHost = serverInstance.getHostname();
+      int port = serverInstance.getGrpcPort();
+      // ensure concurrent request count from the same grpc server cannot exceed maximum
+      if (!_concurrentQueriesCountMap.containsKey(serverHost)) {
+        _concurrentQueriesCountMap.put(serverHost, new AtomicInteger(0));
+      }
+      int concurrentQueryNum = _concurrentQueriesCountMap.get(serverHost).get();
```

Comment:
This is not thread safe
```suggestion
      AtomicInteger concurrentQueryCount = _concurrentQueriesCountMap.computeIfAbsent(serverHost, k -> new AtomicInteger(0));
```

Since we don't have throttling implemented yet, suggest removing the map for now, and add it back when we add the throttling feature. Currently there is no routing layer throttling even for the netty based request handler